contracts/orderbook/Logic/OrderBook.sol

```diff
@@ -0,0 +1,387 @@
+            : order.loanTokenAddress;
+        if (order.timeTillExpiration < block.timestamp) {
+            return false;
+        }
+        if (order.orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+            if (order.loanID != 0 && !_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            uint256 dSwapValue;
+            if (srcToken == order.loanTokenAddress) {
+                amountUsed =
+                    order.loanTokenAmount +
+                    (order.loanTokenAmount * order.leverage) /
+                    10**18; //adjusts leverage
+            } else {
+                amountUsed = queryRateReturn(
+                    order.base,
+                    order.loanTokenAddress,
+                    order.collateralTokenAmount
+                );
+                amountUsed = (amountUsed * order.leverage) / 10**18;
+            }
+            dSwapValue =
+                order.collateralTokenAmount +
+                protocol.getSwapExpectedReturn(
+                    order.loanTokenAddress,
+                    order.base,
+                    amountUsed,
+                    order.loanDataBytes
+                );
+
+            if (order.amountReceived <= dSwapValue) {
+                return true;
+            }
+        } else if (order.orderType == IOrderBook.OrderType.LIMIT_CLOSE) {
+            if (!_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            uint256 dSwapValue;
+            dSwapValue = protocol.getSwapExpectedReturn(
+                order.base,
+                order.loanTokenAddress,
+                order.collateralTokenAmount,
+                order.loanDataBytes
+            );
+            if (order.amountReceived <= dSwapValue) {
+                return true;
+            }
+        } else {
+            if (!_isActiveLoan(order.loanID)) {
+                return false;
+            }
+            bool operand;
+            if (_useOracle[order.trader]) {
+                operand =
+                    order.amountReceived >=
+                    queryRateReturn(
+                        order.base,
+                        order.loanTokenAddress,
+                        order.collateralTokenAmount
+                    );
+            } else {
+                operand =
+                    order.amountReceived >=
+                    getDexRate(
+                        order.base,
+                        order.loanTokenAddress,
+                        order.loanDataBytes,
+                        order.collateralTokenAmount
+                    );
+            }
+            if (operand) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function getDexRate(
+        address srcToken,
+        address destToken,
+        bytes memory payload,
+        uint256 amountIn
+    ) public returns (uint256 rate) {
+        uint256 tradeSize = 10**IERC20Metadata(srcToken).decimals();
```

Comment:
should be this just cashed ? bit its ok for first version

Reply:
I can build some sort of caching so it just does an sload but for this version I think it is fine as is.